pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -77,34 +79,46 @@ func (p *podPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packing, error
 }
 
 // TODO filter instance types based on node contraints like availability zones etc.
-func (p *podPacker) getNodeCapacities() []*nodeCapacity {
+func (p *podPacker) getNodeCapacities(constraints *cloudprovider.Constraints) []*nodeCapacity {
+	p.reserveOnce.Do(
+		func() {
+			for _, nc := range nodeCapacities {
+				kubeletOverhead := binpacking.CalculateKubeletOverhead(nc.total)
+				if ok := nc.reserve(binpacking.MergeResources(constraints.Overhead, kubeletOverhead)); !ok {
+					zap.S().Errorf("Failed to reserve kubelet overhead for node capacity type %v", nc.instanceType)
+				}
+			}
+		})
 	return nodeCapacities
 }
 
+type packingsPerCapacity struct {
+	packed   []*v1.Pod
+	unpacked []*v1.Pod
+}
+
 // packWithLargestPod will try to pack max number of pods with largest pod in
 // pods across all available node capacities. It returns Packing: max pod count
 // that fit; with their node capacities and list of leftover pods
-func (p *podPacker) packWithLargestPod(pods []*v1.Pod) (*Packing, []*v1.Pod) {
+func (p *podPacker) packWithLargestPod(podsToBePacked []*v1.Pod, constraints *cloudprovider.Constraints) (*Packing, []*v1.Pod) {
 	bestPackedPods := []*v1.Pod{}
 	bestCapacities := []*nodeCapacity{}
-	remainingPods := pods
-	// TODO reserve (Kubelet+ daemon sets) overhead for instance types
-	// TODO count number of pods created on an instance type
-	for _, nc := range p.getNodeCapacities() {
+	remainingPods := podsToBePacked
+	for _, nc := range p.getNodeCapacities(constraints) {
 		// check how many pods we can fit with the available capacity
-		packedPods, unpackedPods := p.packPodsForCapacity(nc, pods)
-		if len(packedPods) == 0 {
+		pods := p.packPodsForCapacity(nc, podsToBePacked)
```

Comment:
It's a bit confusing to call something `pods` unless it's a collection of `v1.Pod` structs.